cbmgmtx/ensurebuckethelper.go

```diff
@@ -0,0 +1,104 @@
+package cbmgmtx
+
+import (
+	"context"
+	"errors"
+	"net/http"
+
+	"github.com/couchbase/gocbcorex/cbhttpx"
+	"go.uber.org/zap"
+	"golang.org/x/exp/slices"
+)
+
+type EnsureBucketHelper struct {
+	Logger     *zap.Logger
+	UserAgent  string
+	OnBehalfOf *cbhttpx.OnBehalfOfInfo
+
+	BucketName string
+	BucketUUID string
+
+	confirmedEndpoints []string
+}
+
+type NodeTarget struct {
+	Endpoint string
+	Username string
+	Password string
+}
+
+func (e *EnsureBucketHelper) pollOne(
+	ctx context.Context,
+	httpRoundTripper http.RoundTripper, target NodeTarget,
+) (bool, error) {
+	e.Logger.Debug("polling a single target",
+		zap.String("endpoint", target.Endpoint),
+		zap.String("username", target.Username))
+
+	resp, err := Management{
+		Transport: httpRoundTripper,
+		UserAgent: e.UserAgent,
+		Endpoint:  target.Endpoint,
+		Username:  target.Username,
+		Password:  target.Password,
+	}.GetTerseBucketConfig(ctx, &GetTerseBucketConfigOptions{
+		BucketName: e.BucketName,
+		OnBehalfOf: e.OnBehalfOf,
+	})
+
+	if err != nil {
+		if errors.Is(err, ErrBucketNotFound) {
+			e.Logger.Debug("target responded with bucket not found")
+			return false, nil
+		}
+
+		e.Logger.Debug("target responded with an unexpected error", zap.Error(err))
+		return false, err
+	}
+
+	if e.BucketUUID != "" && resp.UUID != e.BucketUUID {
+		e.Logger.Debug("target responded with success, but the bucket uuid did not match")
+		return false, ErrBucketUuidMismatch
+	}
+
+	e.Logger.Debug("target successfully checked")
+
+	return true, nil
+}
+
+type EnsureBucketPollOptions struct {
+	Transport http.RoundTripper
+	Targets   []NodeTarget
+}
+
+func (e *EnsureBucketHelper) Poll(ctx context.Context, opts *EnsureBucketPollOptions) (bool, error) {
```

Comment:
This only works for created/updated buckets but not deleted?

Reply:
I think this is fine for a 'EnsureBucket' mechanism.  It's use-case is for validating that the bucket is visible to all nodes such that a asking ns_server will give a consistent answer when a memcached node returns 'bucket not found' after creation due to eventual consistency.  The inverse case where a user has delete a bucket gets this behaviour 'naturally' since we get notifications at the memcached level in the form of a dropped connection.